### ISSUE TITLE
Gradient of functions involving `expm`

### ISSUE THREAD

**Issue by danieljtait** [google/jax#2645]
Hi, the following attempt to take the gradient of a function involving a matrix exponential
```
import jax.numpy as jnp
import jax
from jax.scipy.linalg import expm

def foo(mat):
    return jnp.sum(expm(mat))

grad = jax.grad(foo)(jnp.eye(2))
```
is failing with the following assertion error
```
~/anaconda3/envs/envname/lib/python3.7/site-packages/jax/interpreters/ad.py in linearize(traceable, *primals, **kwargs)
     96   pval_primals, pval_tangents = tree_unflatten(out_tree(), out_pvals)
     97   aval_primals, const_primals = unzip2(pval_primals)
---> 98   assert all(aval_primal is None for aval_primal in aval_primals)
     99   if not has_aux:
    100     return const_primals, pval_tangents, jaxpr, consts

AssertionError: 
```

I'm using `jax.__version__  == 0.1.62` and `jaxlib.__version__ == 0.1.43`, thanks!

**Comment by mattjj** [comment]
Thanks for raising this!

Differentiation of expm is a work-in-progress, see #2062.

Hmm should we close this issue as a duplicate, or is it useful to leave open separately?

**Comment by danieljtait** [comment]
My bad I should have found #2062, happy from this to be closed but just to check I'm following the complexities -- the issue is that `expm` is using a `lax.fori_loop` to effectively calculate `matrix_power(A, 2**s)` with `s = f(A)` for some function. 

* It is trying to push the gradient through the `fori_loop` that gives the weird `assert all(...) ` error message
* But even if `fori_loop` was replaced with something out of `jax.experimental.loops` or `scan` or whatever there is going to be an issue with the fact that the range of the loop is determined inside the computation?

Again, happy for this to be deleted, but at a slightly higher level than just the particular case of the `expm` function, what is the canonical way of doing something like this (admittedly silly example but has the same kind of step that is causing an issue)
```
@jax.jit
def foo(A):
    A_L1 = np.linalg.norm(A)
    n = np.maximum(0, np.floor(np.log2(A_L1 / 0.123)))
    def my_body_fun(i, R):
        return np.dot(R, R)
    lower = np.zeros(1, dtype=n.dtype)
    return jax.lax.fori_loop(lower[0], n, my_body_fun, A)
```
I guess my question is; even assuming the mapping `A -> n` is an adaptive step to make sure that `foo(A)` returns the "correct" answer within tolerance of some mathematically differentiable function is it reasonable to expect to 

a) differentiate through an algorithm like this, even if not necessarily efficient to do so 
b) find a, possibly algorithmic implementation of the gradient of `foo` and register that as a `custom_jvp` or as a primitive
c) don't be silly this is a case-by-case issue

**Comment by hawkinsp** [comment]
The forward-mode derivative was merged as part of #2062 and https://github.com/google/jax/issues/3447 is open for the reverse-mode derivative. I'm going to close this issue since I think the remaining work is a duplicate of #3447
